### PR TITLE
fix: restore get_network_name and update RPC API

### LINKED ISSUE #2902: Remove Forest ChainGetName endpoint
**Issue summary**

Forest supports the `ChainGetName` endpoint, returning the network name.

However this method exists only in Forest. An official endpoint achieving similar goal exists and is `Filecoin.StateNetworkName`.

**Task summary**

- [ ] Replace calls to `ChainGetName` with calls to the official endpoint
- [ ] Remove Forest specific endpoint

**Acceptance Criteria**

- [ ] Code that moved from `ChainGetName` to `Filecoin.StateNetworkName` should continue to work

**Other information and links**

https://github.com/filecoin-project/lotus/blob/9321e0fd1404a479d4dee2f1e13d5ea5fbe7ef4e/api/api_full.go#L460

### PR DESCRIPTION
**Pull request by creativcoder**
## Summary of changes

Changes introduced in this pull request:

- Adds a trimmed down version of state_tree_v0 module from `b622af` (without `StateSnapshot` cache) just enough to support `STATE_NETWORK_NAME` RPC API.
- Updates `get_network_name` in state manager to retrieve internal network name from init actor state.
- Removes `chain_get_name` and replaces with `state_network_name`

We need special support for StateTree version 0 to read genesis state trees. These state trees are important and neither FVM2 nor FVM3 supports them.

## Reference issue to close (if applicable)

Closes #2902 

## Other information and links

**Note**: To be able to fully read and write state tree v0 in its entirety, we will need to add StateSnapshot cache support (from `b622af`) and also finish implementing left out methods in the current port.

"

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -181,20 +181,24 @@ walkdir = "2"
 xsalsa20poly1305 = "0.9"
 zstd-safe = "6.0.5"
 
+# hamtv0 specific deps
+forest_hash_utils = "0.1"
```
Comment (lemmih) [inline Cargo.toml:185]: This needs to go.

File: Cargo.toml [hunk 2]
```diff
@@ -181,20 +181,24 @@ walkdir = "2"
...
+serde_bytes = { package = "cs_serde_bytes", version = "0.12" }
```
Comment (lemmih) [inline Cargo.toml:186]: This can stay if there's a good reason to keep it. Write a comment explaining why it cannot be replaced with `fvm_ipld_encoding`.

File: Cargo.toml [hunk 3]
```diff
@@ -181,20 +181,24 @@ walkdir = "2"
...
 criterion = { version = "0.5.1", features = ["async_tokio", "csv"] }
-cs_serde_bytes = "0.12.2"
 fvm3 = { package = "fvm", default-features = false, version = "~3.4", features = ["arb"] }
 fvm_shared3 = { package = "fvm_shared", version = "~3.3", default-features = false, features = ["arb"] }
 libp2p = { version = "=0.51.1", features = ['tcp', 'noise', 'yamux', 'request-response', 'tokio'] }
 multiaddr = "0.17.1"
 num-bigint = { version = "0.4", features = ['quickcheck'] }
 quickcheck = "1"
 quickcheck_macros = "1"
+serde_bytes = { package = "cs_serde_bytes", version = "0.12" }
```
Comment (lemmih) [inline Cargo.toml:201]: Renaming crates is quite confusing. Let's avoid it.

File: src/cli/subcommands/snapshot_cmd.rs [hunk 4]
```diff
@@ -78,8 +79,9 @@ impl SnapshotCommands {
 
                 let epoch = chain_head.epoch();
 
-                let chain_name = chain_get_name((), &config.client.rpc_token)
+                let chain_name = state_network_name((), &config.client.rpc_token)
```
Comment (lemmih) [inline src/cli/subcommands/snapshot_cmd.rs:82]: Do we want `state_network_name` rather than `chain_get_name` for compatibility with Lotus?

File: src/shim/hamtv0/error.rs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    fn from(e: EncodingError) -> Self {
+        Self::Dynamic(anyhow::anyhow!(e))
+    }
+}
+
+impl From<Box<dyn StdError + Send + Sync>> for Error {
+    fn from(e: Box<dyn StdError + Send + Sync>) -> Self {
+        Self::Dynamic(anyhow::anyhow!(e))
+    }
+}
```
Comment (lemmih) [inline src/shim/hamtv0/error.rs:56]: This can be removed.

File: src/shim/hamtv0/hamt.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+#[derive(PartialEq, Eq, Clone, Debug, Serialize_tuple, Deserialize_tuple)]
+pub struct ActorState {
+    /// Link to code for the actor.
+    pub code: Cid,
+    /// Link to the state of the actor.
+    pub state: Cid,
+    /// Sequence of the actor.
+    pub sequence: u64,
+    /// Tokens available to the actor.
+    pub balance: fvm_shared3::econ::TokenAmount,
```
Comment (lemmih) [inline src/shim/hamtv0/hamt.rs:27]:
```suggestion
    pub balance: crate::shim::econ::TokenAmount,
```

File: src/shim/hamtv0/hamt.rs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+use fvm_ipld_encoding::tuple::*;
+use fvm_ipld_encoding::tuple::{Deserialize_tuple, Serialize_tuple};
+use fvm_ipld_encoding::CborStore;
+use serde::{de::DeserializeOwned, Serialize, Serializer};
+use std::borrow::Borrow;
+use std::marker::PhantomData;
+
+/// State of all actor implementations.
+#[derive(PartialEq, Eq, Clone, Debug, Serialize_tuple, Deserialize_tuple)]
+pub struct ActorState {
```
Comment (lemmih) [inline src/shim/hamtv0/hamt.rs:19]: Why is this here? It is not used anywhere. It can be deleted.

File: src/shim/hamtv0/hash_algorithm.rs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+impl HashAlgorithm for Identity {
+    fn hash<X: ?Sized>(key: &X) -> HashedKey
+    where
+        X: Hash,
+    {
+        let mut ident_hasher = IdentityHasher::default();
+        key.hash(&mut ident_hasher);
+        ident_hasher.bz
+    }
+}
```
Comment (lemmih) [inline src/shim/hamtv0/hash_algorithm.rs:81]: This code is unused. Let's delete it.

**Review by lemmih — CHANGES_REQUESTED** [review]
Alright! We're making progress.

# To summarize the situation:

[HAMT](https://ipld.io/specs/advanced-data-layouts/hamt/spec/) is a hashmap implementation on top of [IPLD](https://ipld.io/). There is _no_ specification for HAMT. There's only a prescriptive draft. The HAMT format is _entirely_ determined by Lotus.

The HAMT format is unversioned and has changed over time. We can use the version of the `StateTree` to determine which HAMT format to expect. As of writing, there are 6 `StateTree` versions: 0 through 5. We are using `FVMv2` to parse versions 3 and 4 and `FVMv3` to parse version 5. The exact differences between versions 3, 4, and 5 are not clearly understood and not explicitly documented anywhere.

We want to support all `StateTree` versions, and this PR attempts to add support for version 0. We've discovered that the only HAMT change between version 0 and version 5 is how the `Pointer` structure is serialized.

# Moving forward

The HAMT changes are minor. We should push for the upstream `fvm_ipld_hamt` crate to support multiple HAMT versions, just like we did with `fvm_ipld_amt`.

File: Cargo.toml (hunk 1 above)
Comment (creativcoder) [reply Cargo.toml:185]: Replace with? `forest_hash_utils` is required even for the existing fvm 3.0's hamt impl by `fvm_ipld_hamt` crate (https://github.com/filecoin-project/ref-fvm/blob/3ad5b90f35085e26585e47674844bb72ef09fb06/ipld/hamt/Cargo.toml#L18).

File: src/cli/subcommands/snapshot_cmd.rs (hunk 4 above)
Comment (creativcoder) [reply src/cli/subcommands/snapshot_cmd.rs:82]: Yes, chain_get_name was a non-standard forest specific API.

File: Cargo.toml (hunk 2 above)
Comment (creativcoder) [reply Cargo.toml]: This is not needed for this port, removed.

File: src/shim/hamtv0/error.rs (hunk 5 above)
Comment (creativcoder) [reply src/shim/hamtv0/error.rs]: removed

File: src/shim/hamtv0/hamt.rs (hunk 7 above)
Comment (creativcoder) [reply src/shim/hamtv0/hamt.rs]: Removed

File: src/shim/hamtv0/hash_algorithm.rs (hunk 8 above)
Comment (creativcoder) [reply src/shim/hamtv0/hash_algorithm.rs]: Removed

File: Cargo.toml (hunk 1 above)
Comment (lemmih) [reply Cargo.toml:185]: That is most tragic. The `forest_hash_utils` crate is dead and gone. We either have to upstream a fix or move forward with our own `hamt` crate.

**Comment by creativcoder** [comment]
Blocked until we can switch to fvm 3.5

https://github.com/helix-onchain/filecoin/pull/216

**Comment by lemmih** [comment]
> Blocked until we can switch to fvm 3.5
> 
> [helix-onchain/filecoin#216](https://github.com/helix-onchain/filecoin/pull/216)

Forest now uses FVM-3.5

**Comment by lemmih** [comment]
Converting to draft because CI says the PR is not ready for review.

File: .config/forest.dic [hunk 9]
```diff
@@ -94,3 +94,4 @@ VRF
 WebAssembly
 WebSocket
 zstd
+Sha256
```
Comment (LesnyRumcajs) [inline .config/forest.dic:97]: Perhaps in the docs let's use the longer form, i.e., `SHA-256`. Alternatively, it may be too much hassle and you could just use backticks in the docs themselves.

File: src/shim/state_tree.rs [hunk 10]
```diff
@@ -94,6 +94,7 @@ impl TryFrom<StateTreeVersion> for StateTreeVersionV3 {
 /// Not all the inner methods are implemented, only those that are needed. Feel
 /// free to add those when necessary.
 pub enum StateTree<S> {
+    V0(state_tree_v0::StateTreeV0<Arc<S>>),
```
Comment (LesnyRumcajs) [inline src/shim/state_tree.rs:97]: Could you please elaborate on why it's needed?

File: src/shim/state_tree.rs [hunk 11]
```diff
@@ -179,14 +193,16 @@ where
...
         }
     }
 
     /// Flush state tree and return Cid root.
     pub fn flush(&mut self) -> anyhow::Result<Cid> {
         match self {
-            StateTree::V2(st) => st.flush().map_err(|e| anyhow::anyhow!("{e}")),
+            StateTree::V2(st) => st.flush().map_err(|e| anyhow!("{e}")),
             StateTree::V3(st) => Ok(st.flush()?),
+            _ => todo!(),
```
Comment (LesnyRumcajs) [inline src/shim/state_tree.rs:205]: What do we do with it?

File: src/shim/state_tree.rs [hunk 12]
```diff
@@ -334,3 +351,157 @@ impl From<&ActorState> for ActorStateV2 {
...
+            state: value.state,
+            sequence: value.sequence,
+            balance: value.balance.into(),
+            delegated_address: None,
+        })
+    }
+}
+
+// ported from commit hash b622af
+pub mod state_tree_v0 {
```
Comment (LesnyRumcajs) [inline src/shim/state_tree.rs:368]: Let's move this to a different file.

File: src/shim/state_tree.rs [hunk 13]
```diff
@@ -334,3 +351,157 @@ impl From<&ActorState> for ActorStateV2 {
...
+                .ok_or(anyhow::anyhow!("Init actor address could not be resolved"))?;
+
+            let _state = fil_actor_interface::init::State::load(
+                self.hamt.store(),
+                init_act.code,
+                init_act.state,
+            )?;
+
+            // XXX: can be fixed by adding resolve_method in fil-actor-states
+            todo!("resolve_address method required on init state.")
```
Comment (LesnyRumcajs) [inline src/shim/state_tree.rs:504]: what do we do with it?

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
It'd be great to explain the changes a bit and (most importantly) add tests.

File: src/shim/state_tree.rs [hunk 14]
```diff
@@ -334,3 +351,157 @@ impl From<&ActorState> for ActorStateV2 {
...
+        pub balance: TokenAmount,
+    }
+
+    /// State tree implementation using HAMT. This structure is not thread safe and should only be used
+    /// in sync contexts.
+    pub struct StateTreeV0<S> {
+        hamt: Hamt<S, ActorState>,
+
+        _version: StateTreeVersion,
+        _info: Option<Cid>,
```
Comment (lemmih) [inline src/shim/state_tree.rs:400]: This looks unused and can be removed.

File: src/shim/state_tree.rs [hunk 15]
```diff
@@ -334,3 +351,157 @@ impl From<&ActorState> for ActorStateV2 {
...
+        // /// State cache
+        // snaps: StateSnapshots, // XXX: This is needed when reading writing state tree v0. As present we only implement the
+        // v0 version of state tree usable enough to support the STATE_NETWORK_NAME RPC API.
+    }
+
+    /// Specifies the version of the state tree
+    #[derive(Debug, PartialEq, Clone, Copy, PartialOrd, Serialize_repr, Deserialize_repr)]
+    #[repr(u64)]
+    pub enum StateTreeVersion {
```
Comment (lemmih) [inline src/shim/state_tree.rs:409]: No need to duplicate this here.

File: src/shim/state_tree.rs [hunk 16]
```diff
@@ -94,6 +94,7 @@ impl TryFrom<StateTreeVersion> for StateTreeVersionV3 {
 /// Not all the inner methods are implemented, only those that are needed. Feel
 /// free to add those when necessary.
 pub enum StateTree<S> {
```
Comment (lemmih) [inline src/shim/state_tree.rs:96]: To keep the code consistent, you must distinguish between state-tree and FVM versions. Each FVM state-tree structure covers a range of state-tree versions.

**Review by lemmih — CHANGES_REQUESTED** [review]
Making progress. Some clean-up and documentation is needed.

**Review by lemmih — APPROVED** [review]

File: src/shim/state_tree.rs (hunk 10 above)
Comment (lemmih) [reply src/shim/state_tree.rs]: Done. Wrote which versions are supported by FVM2 and FVM3.

File: src/shim/state_tree.rs (hunk 11 above)
Comment (lemmih) [reply src/shim/state_tree.rs]: Changed it to an error.

File: src/shim/state_tree.rs (hunk 12 above)
Comment (lemmih) [reply src/shim/state_tree.rs]: Done.

File: src/shim/state_tree.rs (hunk 13 above)
Comment (lemmih) [reply src/shim/state_tree.rs]: Changed to an error.

File: src/shim/state_tree_v0.rs [hunk 17]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+// ported from commit hash b622af
+
+// The FVM crates only support state tree versions 3,4 and 5. This module
+// contains read-only support for state tree version 0. This version is required
+// to parse genesis states. Ideally, we would have a library that supports _all_
```
Comment (LesnyRumcajs) [inline src/shim/state_tree_v0.rs:7]: It's not a huge deal, but could we contrive some tests for this? We do have the genesis CAR after all.

**Review by LesnyRumcajs — APPROVED** [review]

File: src/shim/state_tree_v0.rs (hunk 17 above)
Comment (lemmih) [reply src/shim/state_tree_v0.rs:7]: I added two tests for inspecting the state trees of the calibnet genesis block and mainnet genesis block.

Comment (lemmih) [reply src/shim/state_tree_v0.rs:7]: They're in the main `state_tree.rs` file.
